blocks/bricks/__init__.py

```diff
@@ -158,7 +159,56 @@ def _push_initialization_config(self):
                     child.biases_init = self.biases_init
 
 
-class Linear(Initializable):
+class Feedforward(Brick):
+    """Declares an interface for bricks with one input and one output.
+
+    Many bricks have just one input and just one output (activations,
+    :class:`Linear`, :class:`MLP`). To make such bricks interchangable
+    in most contexts they should share an interface for configuring
+    their input and output dimensions. This brick provides such an
+    interface.
+
+    """
+    @abstractproperty
+    def input_dim():
+        """The input dimension of the brick."""
+        pass
+
+    @abstractproperty
+    def output_dim():
+        """The output dimension of the brick."""
+        pass
+
+
+def override_with_attribute(property_, attribute=None):
```

Comment:
Yes, but we're not writing in Java, and Python gives a lot more freedom to the user. I mean, private attributes don't even exist in Python, and to define an abstract class you need to import metaclasses... That kind of shows what Python's intended design is.

In this case I think the readability side definitely weights heavier. For a few reasons:

* This is not about a user passing wrong values or anything, where we want to raise informative errors, but about someone building their own bricks. Someone who is going to build their own bricks can be expected to have read the documentation of the bricks he is inheriting from. We can't babysit them every step along the way. We can even put this `Feedforward` brick in an upcoming "How to write your own brick"-tutorial.
* You are trying to help the user by reminding them to set these attributes. What will happen in reality is that a user who doesn't read the documentation will simply type:
```python
class MyNewBrick(Feedforward):
    def __init__(self, input_dim, output_dim):
        self.input_dim = input_dim
        self.output_dim = output_dim
```
This will crash with `TypeError: Can't instantiate abstract class MyNewBrick with abstract methods input_dim`. Considering we are dealing with someone who was in a rush, and maybe not very good at Python, he won't spend any time on figuring out this error, and simply write:
```python
class MyNewBrick(Brick):
    def __init__(self, input_dim, output_dim):
        self.input_dim = input_dim
        self.output_dim = output_dim
```
The user doesn't care about the fact that he's not using a base class we wrote for him. He just wanted the `input_dim` attribute, and if that's actually getting *more* difficult to set than easier, then he's just going to use something else.
* In general, considering we are dealing with a framework aimed at computer scientists who want to do rapid prototyping and development, I care a lot more about readability.
* Python is based around EAFP and duck-typing. These kind of security checks which are central to Java, C++, etc. are very non-Pythonic.